framework/src/executioners/Transient.C

```diff
@@ -371,25 +316,21 @@ Transient::execute()
   // in the future eventually.
   if (!_app.isRecovering())
     _problem.advanceState();
+  else
+    incrementStepOrReject();
 
   // Start time loop...
-  while (true)
+  while (keepGoing())
```

Comment:
@permcody You will like this. This is the clean loop you wanted long ago ;-). This is in my first commit. It only fails one test in level-set module. I guess our past creeping changes in `Transient` made this easier.